### ISSUE TITLE
Associated type behind a *mut unecessarily results in invariant subtyping

### ISSUE THREAD

**Issue by gmorenz** [rust-lang/rust#68375]
Consider the following code ([playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a73ee6a15b31f56b93fae74c447ea628))

```rust
pub trait Data {
    type Elem;
}

pub struct ViewRepr<A>(A);

impl<'a, A> Data for ViewRepr<&'a A> {
    type Elem = A;
}

pub struct ArrayBase<D: Data> {
    ptr: *mut D::Elem,
    // ptr: *mut f32,
    d: D,
}

pub fn cast_array_view<'shorter, 'longer: 'shorter>(
    input: ArrayBase<ViewRepr<&'longer f32>>
) -> ArrayBase<ViewRepr<&'shorter f32>> {
    input
}
```

`cast_array_view` fails to compile with the first definition of `ptr`, and succeeds with the second, despite the two functions being apparently identical. 

I think the reason this is happening is most likely related to the fact that subtyping of `*mut` [pointers is invariant on the type of their parameter](https://doc.rust-lang.org/nomicon/subtyping.html#variance), and structs containing `*mut T` pointers are invariant with respect to `T`. Since `D::Elem` is being stored behind a `*mut` pointer rustc has decided that `ArrayBase` is invariant with respect to `D`, instead of `D::Elem`.

PS: Thanks to /u/Patryk27 on reddit for looking at this and pointing me in the direction of the variance docs.

**Comment by KrishnaSannasi** [comment]
Using associated types always results in an invariant type bound, so that's where the invaraince is coming from. Also `*mut T` is invariant in `T`, as you noted, but this doesn't make `ArrayBase<D>` invariant in `D`, only invariant in `D::Elem`. But the fact that you used `D::Elem` makes it invariant in `D`.

To get around this:

```
pub trait Data {
    type Elem;
}

pub struct ViewRepr<A>(A);

impl<'a, A> Data for ViewRepr<&'a A> {
    type Elem = A;
}


type ArrayBase<D> = ArrayBaseInner<D, <D as Data>::Elem>;
pub struct ArrayBaseInner<D: Data<Elem = Elem>, Elem> {
    ptr: *mut Elem,
    d: D,
}

pub fn cast_array_view<'shorter, 'longer: 'shorter>(
    input: ArrayBase<ViewRepr<&'longer f32>>
) -> ArrayBase<ViewRepr<&'shorter f32>> {
    input
}
```

**Comment by gmorenz** [comment]
> Using associated types always results in an invariant type bound

Huh, ok. That leaves two questions

 - Why? (For my own education)
 - Is this documented somewhere? Maybe it should be added to the [variance section of the nomicon](https://doc.rust-lang.org/nomicon/subtyping.html#variance)?

**Comment by KrishnaSannasi** [comment]
> Why?

Because there is no general way to figure out the lifetime relationships between the associated type and `Self`. So invariance is the only safe variance.

> Is this documented somewhere? Maybe it should be added to the variance section of the nomicon?

I'm not sure, I just came across this in much the same way you did. It should be added to the nomicon (and the reference while we're at it).

**Comment by KrishnaSannasi** [comment]
Oh, you don't need the type alias. This works

```rust
pub struct ArrayBase<D: Data<Elem = Elem>, Elem = <D as Data>::Elem> {
    ptr: *mut Elem,
    d: D,
}
```

**Comment by gmorenz** [comment]
I tend to agree with your conclusion that that suggests it would be possible to do better.

Actually trying to use that in real code (ndarray) results in me running into a different issue, which I believe is a version of #28895. I've added a minimized example to that issue.
